prompt/prompt-tokenizer/src/commonMain/kotlin/ai/koog/prompt/tokenizer/tiktoken/TiktokenEncoder.kt

```diff
@@ -0,0 +1,117 @@
+package ai.koog.prompt.tokenizer.tiktoken
+
+import ai.koog.prompt.tokenizer.Tokenizer
+
+/**
+ * A tokenizer implementation that uses a token encoding vocabulary and a regex pattern to tokenize
+ * text into a series of token IDs. The tokenization process utilizes byte pair encoding (BPE) for
+ * segments of text not directly found in the vocabulary.
+ *
+ * This implementation is based on the reference Rust implementation from
+ * https://github.com/openai/tiktoken/blob/main/src/lib.rs.
+ *
+ * @constructor Initializes the TiktokenEncoder with the specified vocabulary, regex pattern for
+ * matching tokens, and a token ID for unknown tokens.
+ * @param vocabulary A mapping of ByteArrayKey to token ID that represents the token encoding vocabulary.
+ * @param pattern A regular expression used to match text patterns for initial token matching.
+ * @param unkTokenId The token ID used for unknown tokens not present in the vocabulary.
+ */
+public open class TiktokenEncoder(
```

Comment:
Do we really want it open?

Reply:
Yes, I didn't add implementations for it because that is quite out of scope of this MR. There are multiple versions of TiktokenEncoder that we may add in the future